Adding aggregated services to catalog, topology and various views.

wtrocki opened the pull request:
## Motivation

This PR adds Firepig OpenShift Community plugin that will put base for discovering services in the OpenShift console. This PR is still containing some elements not aligned with the design. There are also couple issues that we need to solve. Creating for early preview.

## Verification

To verify changes we need to install firepig operator 
The operator can be added as a separate catalog source.
Currently, we offer only alpha stream that represents the latest master changes.

### Operator installation Procedure

1. Create catalog.yml file
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: rhoas-operators
  namespace: openshift-marketplace
spec:
  displayName: RHOAS Operators
  icon:
    base64data: ""
    mediatype: ""
  image: quay.io/rhoas/service-operator-registry:autolatest
  priority: -400
  publisher: RHOAS
  sourceType: grpc
  updateStrategy:
    registryPoll:
      interval: 260s
```

1. Ad catalog source

```
oc apply -f catalog.yml
```

1. Wait 1 minute and go to openshift catalog. 
Type rhoas and install operator.

### Verification procedure

1. Go to the Add screen and look for `Managed Service section`.
2. Follow instructions on the wizard until redirected to the topology screen.

## Notes

1. Discovered services are now static (we do not have dynamic capabilities to populate the catalog yet). 
We can create follow up issues for solving those

2. Icons are still temporary

3. Designs are still in progress. 

4. There are still some issues with topology that we would love to get help

5. Catalog integration needs improvements

## Progress

- [x] Base for the plugin 
- [x] Initial design in topology
- [x] Catalog plugin
- [x] Error handling
- [x] Applying proper designs (currently in works)
- [x] Service binding connections (currently hacked - no visuals)

wtrocki commented on frontend/packages/rhoas-plugin/locales/ja/rhoas-plugin.json:
```diff
@@ -0,0 +1,37 @@
+{
```
Do we need to provide translations? How to deal with those? Leave those empty?

jeff-phillips-18 replied:
Do not create the other locales, they get created later.

christiemolloy replied:
+1

wtrocki commented on frontend/packages/rhoas-plugin/api-mock/README.md:
```diff
@@ -0,0 +1,7 @@
+## Mocking capability for RHOAS Operator
```
This will be removed at some point - We have that for demo purposes and to simulate errors.

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/createConnector.ts:
```diff
@@ -0,0 +1,22 @@
+
+export const getCreateConnector = (createHints: string[]) => {
+  if (
+    createHints
```
This is disabled because I could not find reliable way to find hints. 
SBO integration works but generally arrow is not drawn for this

jeff-phillips-18 replied:
Rather than using the default `nodeDropTargetSpec` you need to create your own and set the dropHint there.
See `OperatorBackedServcie.tsx`

jeff-phillips-18 replied:
SBRs are not showing up because they are only crated by the operators plugin which only looks for connections to the OBS groups. You will need to do something similar in your data model transformer. See `operators-data-transformer.ts` for an example.

wtrocki replied:
I have done this but ended up having to move so many things that really made 2 places in the console that define SBO based connections.

wtrocki commented:
FYI @jeff-phillips-18 @christianvogt

invincibleJai commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,160 @@
+    return (
+      <Flex direction={{ default: 'column' }}>
+        <FlexItem>
+          RHOAS can include Managed Kafka, Service Registry, custom resources for Managed Kafka, and
+          Open Data Hub.
```
we may need to translate this as well?

christiemolloy replied:
+1

invincibleJai commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,160 @@
+    managedServicesCard,
+    tokenSecret,
+  ]);
+  return [services, true, undefined];
```
just to be sure loaded will be always `true`?

invincibleJai commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,160 @@
+    },
+  ];
+
+  const services = React.useMemo(() => (tokenSecret ? managedKafkaCard : managedServicesCard), [
```
shall we add if `tokenSecret` is loaded , exists and has no error?

christiemolloy replied:
this will be coming soon when we add error handling

invincibleJai commented on frontend/packages/rhoas-plugin/src/components/managed-services-kafka/ManagedKafkas.tsx:
```diff
@@ -0,0 +1,97 @@
+    kind: referenceForModel(ManagedKafkaRequestModel),
+    name: ManagedKafkaRequestCRName,
+    namespace: currentNamespace,
+    isList: false
```
can add add optional as well 

```
...
isList: false
optional: true
```

christiemolloy replied:
+1

invincibleJai commented on frontend/packages/rhoas-plugin/src/components/managed-services-kafka/ManagedKafkas.tsx:
```diff
@@ -0,0 +1,97 @@
+    createKafkaRequestFlow()
+  }, []);
+
+  const [watchedKafkaRequest] = useK8sWatchResource<KafkaRequest>({
```
we don't need loaded , error state here along with `watchedKafkaRequest`?

christiemolloy replied:
this will be coming soon when we add error handling

invincibleJai commented on frontend/packages/rhoas-plugin/src/components/managed-services-list/ManagedServicesList.tsx:
```diff
@@ -0,0 +1,77 @@
+import { Gallery, GalleryItem, Label } from '@patternfly/react-core';
+import { CatalogTile } from '@patternfly/react-catalog-view-extension';
+import { history } from '@console/internal/components/utils';
+// import AccessManagedServices from '../access-managed-services/AccessManagedServices';
```
nit: can remove comments

christiemolloy replied:
+1

invincibleJai commented on frontend/packages/rhoas-plugin/src/components/managed-services-list/ManagedServicesList.tsx:
```diff
@@ -0,0 +1,77 @@
+    );
+  };
+
+  const defaultHintBlockText = `Select Managed Service you would like to connect with. To use the Red Hat Managed Services you need to have account and at least one active service in https://cloud.redhat.com`;
```
we may need translation

christiemolloy replied:
+1

invincibleJai commented on frontend/packages/rhoas-plugin/src/components/streams-list/StreamsInstanceTable.tsx:
```diff
@@ -0,0 +1,95 @@
+  const { t } = useTranslation();
+
+  const formatTableRowData = () => {
+    const tableRow = [];
+    kafkaArray.forEach(row => {
+      const { name, bootstrapServerHost, provider, region, owner, createdAt } = row;
+      tableRow.push({
+        cells: [
+          { title: name },
+          { title: <a href="/">{bootstrapServerHost}</a> },
+          { title: provider },
+          { title: region },
+          { title: <a href="/">{owner}</a> },
+          { title: <Timestamp timestamp={createdAt} /> },
+        ]
+      });
+    });
+
+    kafkaArray.forEach((kafka, index) => {
+      if (currentKafkaConnections.includes(kafka.id)) {
+        tableRow[index].disableSelection = true;
+      }
+    })
```
Can we make use of map as below

```suggestion
   const tableRow  = kafkaArray.map(({ kafka , name, bootstrapServerHost, provider, region, owner, createdAt }) => {
   return {
        cells: [
          { title: name },
          { title: <a href="/">{bootstrapServerHost}</a> },
          { title: provider },
          { title: region },
          { title: <a href="/">{owner}</a> },
          { title: <Timestamp timestamp={createdAt} /> },
        ],
       ...(currentKafkaConnections.includes(kafka.id) && { disableSelection : true }
      }
})
```

christiemolloy replied:
+1

invincibleJai commented on frontend/packages/rhoas-plugin/src/models/rhoas.ts:
```diff
@@ -0,0 +1,41 @@
+import { K8sKind } from '@console/internal/module/k8s';
+
+export const ManagedKafkaRequestModel: K8sKind = {
+  apiGroup: 'rhoas.redhat.com',
+  apiVersion: 'v1alpha1',
```
nit: would be good to move `rhoas.redhat.com` and `v1alpha1` as const and use below as well, WDYT?

christiemolloy replied:
Great idea!

christiemolloy replied:
+1

invincibleJai reviewed (commented):
Thanks, have added some comments PTAL if it make sense and  will verify it next

invincibleJai commented:
/ok-to-tes

invincibleJai commented:
/ok-to-test

jeff-phillips-18 commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,150 @@
+  );
+};
+
+export const ConnectedTopologyHelmReleasePanel: React.FC<TopologyHelmReleasePanelProps> = ({
```
Copy/Paste error?

christiemolloy replied:
+1

jeff-phillips-18 commented on frontend/packages/rhoas-plugin/src/topology/rhoas-topology-plugin.tsx:
```diff
@@ -0,0 +1,69 @@
+    type: 'Topology/CreateConnector',
+    properties: {
+      getCreateConnector: getExecutableCodeRef(() =>
+        import('./createConnector' /* webpackChunkName: "operators-service-bindings" */).then(
```
Copy/Paste error

christianvogt replied:
This was marked as resolved but the change wasn't made. The chunk name is incorrect.

jeff-phillips-18 commented on frontend/packages/topology/src/components/side-bar/TopologySideBar.tsx:
```diff
@@ -73,6 +74,9 @@ export const SelectedItemDetails: React.FC<SelectedItemDetailsProps> = observer(
       if (selectedEntity.getType() === TYPE_HELM_RELEASE) {
         return <TopologyHelmReleasePanel helmRelease={selectedEntity} />;
       }
+      if (selectedEntity.getType() === "ManagedKafkaConnection") {
```
Constant please

christiemolloy replied:
+1

wtrocki commented:
Thank you so much for review. We going to apply all suggestions and also add valid texts and other changes needed.

christiemolloy commented on frontend/packages/rhoas-plugin/src/actions/modify-kafka-connection.tsx:
```diff
@@ -0,0 +1,16 @@
+import { k8sKill } from '@console/internal/module/k8s';
+import { ManagedKafkaConnectionModel } from '../models/rhoas';
+
+export const deleteManagedKafkaConnection = (name: string, namespace: string) => {
```
+1

christiemolloy commented on frontend/packages/rhoas-plugin/src/actions/modify-kafka-connection.tsx:
```diff
@@ -0,0 +1,16 @@
+
+export const deleteManagedKafkaConnection = (name: string, namespace: string) => {
+  return {
+    labelKey: 'rhoas-plugin~Delete Kafka Connection',
```
+1

christiemolloy commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,160 @@
+  TextVariants,
+} from '@patternfly/react-core';
+import { LockIcon } from '@patternfly/react-icons';
+import { AccessTokenSecretName, managedKafkaIcon, operatorIcon } from '../../const';
+import { useTranslation } from 'react-i18next';
```
+1

invincibleJai commented:
> Thank you so much for review. We going to apply all suggestions and also add valid texts and other changes needed.

Thanks, i tried to verify this as well. Few pointers on that

- `CatalogSource` name should start with small case looks like i.e `firepig` not `Firepig`
-  i could install the operator and in dev Perspective could see Card under Add , went on catalog. 🎉 
- In catalog until Api token is not added, can see button without text (maybe we need to remove it)

![image](https://user-images.githubusercontent.com/5129024/108503702-d5c7a500-72da-11eb-8aa6-c8422074dbf1.png)

- after i added `token` can see button `Connect` but on clicking connect , UI breaks

![image](https://user-images.githubusercontent.com/5129024/108503851-0c9dbb00-72db-11eb-97ad-071a8beaf8c1.png)

in network can see `404` for api/kubernetes/apis/rhoas.redhat.com/v1alpha1/namespaces/jai-test/managedkafkarequests/rh-managed-services-kafka-request

wtrocki commented:
> in network can see 404 for api/kubernetes/apis/rhoas.redhat.com/v1alpha1/namespaces/jai-test/managedkafkarequests/rh-managed-services-kafka-request 

Thank you so much for review!
We have done change in Operator API that renamed resource. 
Going to update the operator with all the fixes next week and request a review.

wtrocki commented:
Updated Operator to new official catalog source. 
We hoping to push UI tweaks this weekend and shared some document with outstanding issues that possibly could become jiras.

wtrocki commented:
// Update

I have updated PR acording to designs. Squashed many commits and verified entire design. 
The only outstanding element is Topology SBO integration. 

Please note that this version will work only with operator 0.3.0 (we have been asked to made breaking changes in the API

wtrocki commented:
I do not understand why build is failing.
Any insights will be appreciated. 
Code is already meeting all standards and passes the local verification

```
{"component":"clonerefs","error":"1 clone records failed","file":"prow/cmd/clonerefs/main.go:40","func":"main.main","level":"fatal","msg":"Failed to clone refs","severity":"fatal","time":"2021-03-12T15:12:47Z"}
error: build error: error building at STEP "RUN umask 0002 && /clonerefs && find /go/src -type d -not -perm -0775 | xargs --max-procs 10 --max-args 100 --no-run-if-empty chmod g+xw": error while running runtime: exit status 1
```

wtrocki commented:
Bumped version to see if builds will pass.

wtrocki commented:
@invincibleJai @jeff-phillips-18 @christianvogt this is rebased. Validated and fully functional - I have updated to latest patterns on master. Ready for review and approval.

wtrocki commented:
Who should I ping for CI/CD issues?

Getting:
```
{"component":"clonerefs","error":"1 clone records failed","file":"prow/cmd/clonerefs/main.go:40","func":"main.main","level":"fatal","msg":"Failed to clone refs","severity":"fatal","time":"2021-03-18T17:54:53Z"}
error: build error: error building at STEP "RUN umask 0002 && /clonerefs && find /go/src -type d -not -perm -0775 | xargs --max-procs 10 --max-args 100 --no-run-if-empty chmod g+xw": error while running runtime: exit status 1
```

This is has nothing to do with my code.

spadgett commented:
> Who should I ping for CI/CD issues?
> 
> Getting:
> 
> ```
> {"component":"clonerefs","error":"1 clone records failed","file":"prow/cmd/clonerefs/main.go:40","func":"main.main","level":"fatal","msg":"Failed to clone refs","severity":"fatal","time":"2021-03-18T17:54:53Z"}
> error: build error: error building at STEP "RUN umask 0002 && /clonerefs && find /go/src -type d -not -perm -0775 | xargs --max-procs 10 --max-args 100 --no-run-if-empty chmod g+xw": error while running runtime: exit status 1
> ```
> 
> This is has nothing to do with my code.

I think this is related to your PR. Did you mean to add a Git submodule?

wtrocki commented:
@spadgett good catch. We had submodule done for temporary testing and mocking but it should be removed from this PR. That will be easy fix.

wtrocki commented:
/test e2e-gcp-console

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+import { SecretModel } from '@console/internal/models';
+import * as UIActions from '@console/internal/actions/ui';
+import './TopologyKafkaPanel.css';
+// import { KafkaConnectionModel } from '../../models';
```
Remove this

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+      <div className="overview__sidebar-pane-head resource-overview__heading">
+        <h1 className="co-m-pane__heading">
+          <div className="co-m-pane__name co-resource-item">
+            {/* <ResourceIcon
```
Remove

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+              actionClose={<AlertActionCloseButton onClick={handleAlertFunction} />}
+              isInline
+            >
+              This resource represents service that exist outside your cluster. To view details
```
This needs to be localized

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+              This resource represents service that exist outside your cluster. To view details
+              about resource please go to
+              <a href="https://cloud.redhat.com/beta/application-services/openshift-streams/">
+                {' '}
```
Remove those

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/createConnector.ts:
```diff
@@ -0,0 +1,22 @@
+export const getCreateConnector = (createHints: string[]) => {
+  if (
+    createHints
+    // Enabling hints and drop targets requires copying
```
This can be removed. It is outdated

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/TrapezoidBaseNode.scss:
```diff
@@ -0,0 +1,63 @@
+@import '../../../../topology/src/components/topology-utils.scss';
+
+.odc-trapezoid-node {
```
We should not use our own prefixes. Use rhoas

christianvogt replied:
```suggestion
.rhoas-trapezoid-base-node {
```

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.css:
```diff
@@ -0,0 +1,3 @@
+.kafka-panel-alert {
```
We need to have common parent for all CSS. RHOAS

wtrocki replied:
rhoas-topology-kafka-panel

wtrocki commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,147 @@
+    <Flex direction={{ default: 'column' }}>
+      <FlexItem>
+        <TextContent>
+          <Text component={TextVariants.p}>TO DO: Add description</Text>
```
We need to swap strings

wtrocki commented on frontend/packages/rhoas-plugin/src/utils/k8sUtils.ts:
```diff
@@ -0,0 +1,52 @@
+ * @param {number} timeoutInMs
+ * @returns {Promise<K8sResourceKind>}
+ */
+export const k8sWaitForUpdate = (kind, resource, checkCondition, timeoutInMs) => {
```
This can be updated in the source.

invincibleJai commented:
Have verified flow locally, can see 
- was able to unlock 
- list kafka instance and create
- on creation can see `KC` 

![image](https://user-images.githubusercontent.com/5129024/112027049-d7e08600-8b5c-11eb-8e6f-b9a753cccf56.png)


Observation: can have other options (common context menu) as well as we have on other workload and side bar should have the name and click on it should redirect the user to details page

wtrocki commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,147 @@
+    },
+  ];
+
+  const cloudServicesCard: CatalogItem[] = [
```
This should be part of the use memo

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+
+type TopologyRhoasPanelProps = PropsFromState & PropsFromDispatch & OwnProps;
+
+const DetailsComponent: React.FC<any> = ({ obj }) => {
```
Should be a separate component file for better separation.

christianvogt replied:
and don't use `any` type.

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+  );
+};
+
+export const ConnectedTopologyRhoasPanel: React.FC<TopologyRhoasPanelProps> = ({
```
Should be a separate component file for better separation.

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+  item,
+  selectedDetailsTab,
+  onClickTab,
+}: TopologyRhoasPanelProps) => {
```
```suggestion
}) => {
```

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+  );
+};
+
+const ResourcesComponent = ({ obj }) => {
```
Missing props:
```suggestion
const ResourcesComponent: React.FC<ResourcesComponentProps> = ({ obj }) => {
```

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,163 @@
+            >
+              This resource represents service that exist outside your cluster. To view details
+              about resource please go to
+              <a href="https://cloud.redhat.com/beta/application-services/openshift-streams/">
+                {' '}
+                OpenShift Streams Apache Kafka{' '}
+              </a>{' '}
```
Do we need the external url icon for such links? Should probably also use the PF `<Button variant="link"` component.
```suggestion
              {' '}
              <a href="https://cloud.redhat.com/beta/application-services/openshift-streams/" target="_blank">
                OpenShift Streams Apache Kafka
              </a>{' '}
```

christianvogt commented on frontend/packages/rhoas-plugin/src/utils/k8sUtils.ts:
```diff
@@ -0,0 +1,52 @@
+ * @param {number} timeoutInMs
+ * @returns {Promise<K8sResourceKind>}
+ */
+export const k8sWaitForUpdate = (kind, resource, checkCondition, timeoutInMs) => {
```
update core util instead of making a copy

christianvogt replied:
This comment was marked as resolved. We also discussed this on our call.
However no change has been made.

christianvogt commented on frontend/packages/rhoas-plugin/src/utils/resourceCreators.ts:
```diff
@@ -0,0 +1,257 @@
+  const localArray = [];
+  const kafkaConnections = await k8sGet(KafkaConnectionModel, null, currentNamespace);
+  if (kafkaConnections) {
+    const callback = (kafka) => {
+      const { kafkaId } = kafka.spec;
+      localArray.push(kafkaId);
+    };
+    kafkaConnections.items.map(callback);
+    return localArray;
```
```suggestion
    return kafkaConnections.items.map((kafka) => kafka.spec.kafkaId);
```

wtrocki commented on frontend/.vscode/settings.json:
```diff
@@ -70,7 +70,8 @@
   "cucumberautocomplete.strictGherkinCompletion": true,
-  "editor.quickSuggestions": true
+  "editor.quickSuggestions": true,
+  "typescript.format.insertSpaceAfterFunctionKeywordForAnonymousFunctions": false
```
There is TODO bellow for supporting prettier. 
When working with Vscode I got some strange formatting issues every time I save. 
Prettier plugin writes stuff but then vscode sometimes race condition as well. Having this option set can resolve it but this is also not target of this PR so I will remove it and put it iny my private settings

wtrocki replied:
Lets remove it

christiemolloy replied:
Done

wtrocki commented on frontend/packages/rhoas-plugin/src/actions/delete-kafka-connection.tsx:
```diff
@@ -0,0 +1,17 @@
+import { KebabOption } from '@console/internal/components/utils';
+import { KafkaConnectionModel } from '../models/rhoas';
+
+export const deleteKafkaConnection = (name: string, namespace: string): KebabOption => {
```
Yes. Resolved that already :)

invincibleJai commented on frontend/packages/rhoas-plugin/src/models/rhoas.ts:
```diff
@@ -0,0 +1,41 @@
+  plural: 'kafkaconnections',
+  label: 'Kafka Connection',
+  labelPlural: 'Kafka Connections',
+  abbr: 'AKC',
```
why `AKC` , shouldn't this be `KC`

wtrocki replied:
KC is used for cluster defitions, keycloak and couple other reaources. Conflicts were almost guaranteed so we added Apache to it

christiemolloy commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceInstance.tsx:
```diff
@@ -0,0 +1,103 @@
+  disableCreateButton,
+}: ServiceInstanceProps) => {
+  const [allKafkasConnected, setAllKafkasConnected] = React.useState<boolean>(false);
+  const [textInputNameValue, setTextInputNameValue] = React.useState<string>('');
+  const [pageKafkas, setPageKafkas] = React.useState<CloudKafka[]>(kafkaArray);
+  const { t } = useTranslation();
+
+  React.useEffect(() => {
+    setPageKafkas(kafkaArray);
+  }, [kafkaArray]);
+
+  const handleTextInputNameChange = (value: string) => {
+    const filteredKafkas = kafkaArray.filter((kafka) => kafka.name.includes(value));
+    setPageKafkas(filteredKafkas);
+    setTextInputNameValue(value);
+  };
```
Done

christiemolloy commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceInstance.tsx:
```diff
@@ -0,0 +1,103 @@
+                <StreamsInstanceTable
+                  kafkaArray={kafkaArray}
+                  pageKafkas={pageKafkas}
+                  handleTextInputNameChange={handleTextInputNameChange}
```
Done

christiemolloy commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceListPage.tsx:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  if (!watchedKafkaRequest || !watchedKafkaRequest.status) {
+    return (
+      <>
+        <LoadingBox />
+      </>
+    );
```
Done

christiemolloy commented on frontend/packages/rhoas-plugin/src/topology/components/TrapezoidBaseNode.tsx:
```diff
@@ -0,0 +1,184 @@
+
+  return (
+    <Tooltip
+      content={t('topology~Create a binding connector')}
```
Done

christiemolloy commented on frontend/packages/rhoas-plugin/package.json:
```diff
@@ -0,0 +1,19 @@
+    "entry": "src/plugin.ts",
+    "exposedModules": {
+      "constants": "src/const.ts",
+      "catalog": "src/catalog"
```
Done

christiemolloy commented on frontend/packages/rhoas-plugin/src/utils/resourceCreators.ts:
```diff
@@ -0,0 +1,257 @@
+ * @param kafkaName
+ * @param currentNamespace
+ */
+export const deleteKafkaConnection = async (kafkaName: string, currentNamespace: string) => {
```
Done

christiemolloy commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,147 @@
+import { ServiceAccountCRName, kafkaIcon, operatorIcon } from '../../const';
+import { CloudServiceAccountRequest } from '../../models';
+import { isResourceStatusSuccessfull } from '../../utils/conditionHandler';
+import { referenceForModel } from '@console/internal/module/k8s';
```
Done

wtrocki commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,147 @@
+        <FlexItem>{token}</FlexItem>
+      </Flex>
+    );
+  };
```
We still see this in designs. Will check to make sure

christiemolloy commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,147 @@
+        href: '',
+      },
+      details: {
+        properties: [{ label: 'Type', value: 'Red Hat Cloud Service' }],
```
Do you mean to remove `label: 'Type'` ? im getting this TS error: `Property 'label' is missing in type '{ value: string; }' but required in type 'CatalogItemDetailsProperty'`

christianvogt replied:
Specifically I was wondering why the need for this properties entry at all? What value does it provide for the user.

wtrocki replied:
I thought this can be used to add categories based on helm having something similar. PLOT TWIST it is not useful for it so we can delete it

christianvogt replied:
Helm has a facet filter for helm repositories and therefore provides properties which tells the user which repo it belongs to.

In this case, having a property telling the user that this card is named `Red Hat Cloud Service` (as per the title) and is of `type` `Red Hat Cloud Service` seems a bit redundant.

wtrocki commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,147 @@
+      attributes: {
+        version: '1',
+        type: 'kafka',
+      },
```
I felt that this will give us ability to do filter things (helm does similar things with chartType`

wtrocki replied:
Let's remove it

wtrocki replied:
https://github.com/bf2fc6cc711aee1a0c2a/openshift-console-plugin/blob/main/frontend/packages/rhoas-plugin/console-extensions.json#L23-L26

christiemolloy commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceInstance.tsx:
```diff
@@ -0,0 +1,103 @@
+  kafkaArray: CloudKafka[];
+  selectedKafka: number;
+  setSelectedKafka: (selectedKafka: number) => void;
+  currentKafkaConnections: string[];
```
Not sure if you're suggesting to replace the wording here or pass in the raw `kafkaConnections` value instead. will need to take a look further

wtrocki replied:
This links to the other suggestion above.

wtrocki commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceListPage.tsx:
```diff
@@ -0,0 +1,137 @@
+  const { t } = useTranslation();
+  const [kafkaCreateError, setKafkaCreateError] = React.useState<string>();
+
+  React.useEffect(() => {
+    const createKafkaRequestFlow = async () => {
+      await createCloudServicesRequestIfNeeded(currentNamespace);
+
+      const currentKafka = await listOfCurrentKafkaConnectionsById(currentNamespace);
+      if (currentKafka) {
+        setCurrentKafkaConnections(currentKafka);
+      }
+    };
+    createKafkaRequestFlow();
+  }, [currentNamespace]);
```
Operator doesn't provide refresh capabilities because of the API limitations. Once we get results they will stay the same until we patch existing request.

wtrocki commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceListPage.tsx:
```diff
@@ -0,0 +1,137 @@
+    );
+  }
+
+  if (!watchedKafkaRequest || !watchedKafkaRequest.status) {
```
In this situation (specifics to how Java Operator deals with resources) loaded will be true and error will be null but our operator can still take long time to read API.

wtrocki commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceListPage.tsx:
```diff
@@ -0,0 +1,137 @@
+    if (!isAcccesTokenSecretValid(watchedKafkaRequest)) {
+      return (
+        <ServicesErrorState
+          title={t('Could not fetch services')}
+          message={t('Could not connect to RHOAS with API Token')}
```
WOW. I do not know how but this worked without prefix.

wtrocki commented on frontend/packages/rhoas-plugin/src/plugin.ts:
```diff
@@ -0,0 +1,62 @@
+      models: _.values(models),
+    },
+  },
+  {
+    type: 'FeatureFlag/Model',
+    properties: {
+      model: models.KafkaConnectionModel,
+      flag: FLAG_RHOAS_KAFKA,
+    },
+  },
+  {
+    type: 'FeatureFlag/Model',
+    properties: {
+      model: models.CloudServiceAccountRequest,
+      flag: FLAG_RHOAS,
+    },
+  },
```
No problem. I do not see helm having flags in the json file yet and had problems when moving them. Happy to discuss.

christianvogt replied:
This should work as an example:
```
  {
    "type": "console.flag/model",
    "properties": {
      "flag": "RHOAS_FLAG",
      "model": {
        "group": "rhoas.redhat.com",
        "version": "v1alpha1",
        "kind": "CloudServiceAccountRequest"
      }
    }
  }
```

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/KafkaNode.scss:
```diff
@@ -0,0 +1,3 @@
+.co-m-resource-kafka-connection {
+  background-color: --pf-global--palette--red-300;
+}
```
This is somewhat some automatic feature/magic that puts colors on AKC labels - if it remove that things  

So pattern for those is `.co-m-resource-{{name-of-your-crd}}` but I'm not sure if there is some flag to avoid this

christianvogt replied:
Please add a comment to this please.
I know we plan to improve the way we assign colors with a future extetnsion.

wtrocki commented on frontend/packages/rhoas-plugin/src/topology/components/KafkaNode.tsx:
```diff
@@ -0,0 +1,71 @@
+  const iconRadius = Math.min(width, height) * 0.25;
+  const { radius } = calculateRadius(size);
+  const spec = React.useMemo(() => obsDropTargetSpec(serviceBinding), [serviceBinding]);
+  const [dndDropProps, dndDropRef] = useDndDrop(spec, { element, ...props });
```
For this I would need help of @jeff-phillips-18 . Can we have all props passed?

christianvogt replied:
The additional props should only be passed if the the `spec` requires them.
In this case it looks like the spec only requires `element` as an additional prop. 

We really should look at providing some common spec or simply writing one for this particular use case so that we don't reuse a spec that was meant for the operator backed service nodes. cc @jeff-phillips-18

jeff-phillips-18 replied:
Agreed. Reused this for now since we will be adding extensions to be able to create edges.

wtrocki replied:
Are there any actions needed from us here?

christianvogt replied:
I thought you only need to provide `element`. But I suppose it's no big deal since the rest are simply ignored and unused.

wtrocki commented:
Almost all comments applied and resolved (just 4 open questions/issues left out of 82).
Ready for another verification/review

christiemolloy commented:
@christianvogt there was a comment you left about adding a file extension in the package.json file in rhoas-plugin. When I added `.tsx` as per your suggestion, the build of frontend fails. the other plugins seem to just be referencing `catalog` instead of `catalog.tsx`.. what do you think?

<img width="280" alt="Screen Shot 2021-03-24 at 5 45 45 AM" src="https://user-images.githubusercontent.com/20118816/112289247-84595f80-8c64-11eb-95f9-9c9be25bbd84.png">

wtrocki commented:
@christiemolloy I think I skipped that, but we can remove const.ts as it is not used at the moment

wtrocki commented:
Fixed minor build issues. This should be ready to go. Our team done another round of testing. 
We also rebased this PR with master and applied required fixes.

christianvogt commented:
> @christianvogt there was a comment you left about adding a file extension in the package.json file in rhoas-plugin. When I added `.tsx` as per your suggestion, the build of frontend fails. the other plugins seem to just be referencing `catalog` instead of `catalog.tsx`.. what do you think?
> 
> <img alt="Screen Shot 2021-03-24 at 5 45 45 AM" width="280" src="https://user-images.githubusercontent.com/20118816/112289247-84595f80-8c64-11eb-95f9-9c9be25bbd84.png">

@christiemolloy I was only suggesting it for consistency. Maybe we should be consistent and instead omit the extension :)  Do what works for now. We can sweep all packages for better consistency in the future.

cc @vojtechszocs thoughts? and why would it cause the build to fail. Maybe we sweep all packages and remove the extension of exposed modules?

christianvogt commented on frontend/packages/dev-console/integration-tests/support/pageObjects/add-flow-po.ts:
```diff
@@ -96,6 +96,7 @@ export const catalogPO = {
     builderImage: 'ul:nth-child(3) > li:nth-child(1) > a', // This needs to be changed
     template: 'a[href="/?catalogType=Template"]',
     serviceClass: '[data-test="kind-cluster-service-class"]',
+    managedServices: '[data-test="kind-managed-service"]',
```
@makambalaji ^^

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/ResourceComponent.tsx:
```diff
@@ -0,0 +1,29 @@
+
+  return (
+    <ul>
+      <h3>Secret</h3>
```
Missing localization.

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,92 @@
+import { DetailsComponent } from './DetailsComponent';
+
+type PropsFromState = {
+  selectedDetailsTab?: any;
```
Isn't this of type `string` ? or `string` ?
We don't want to use the `any` type.

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,92 @@
+  // Resource
+  const akc = item?.getData().resource;
+  if (!akc) {
+    return <>No data</>;
```
Localize.
```suggestion
    return t('rhoas-plugin~No data');
```

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,92 @@
+    return <>No data</>;
+  }
+
+  const handleAlertFunction = () => {
+    if (showAlert) {
+      setShowAlert(false);
+    }
+  };
```
Redundant check since this function is only called if `showAlert` is truthy.
```suggestion
  const handleAlertFunction = () => {
      setShowAlert(false);
  };
```

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,92 @@
+        selectedTab={selectedDetailsTab}
+        onClickTab={onClickTab}
+        tabs={[
+          { name: 'Details', component: navFactory.details(DetailsComponent).component },
+          { name: 'Resources', component: ResourcesComponent },
```
Missing localization.

christianvogt commented on frontend/packages/rhoas-plugin/src/components/service-table/ServiceInstanceTable.css:
```diff
@@ -0,0 +1,3 @@
+.rhoas-plugin--service-table tr>*:empty {
```
Convention is to use each component name part in the css class name.
`rhoas-plugin` is a bit long. Maybe shorten to `rhoas` instead. Furthermore there is inconsistencies in this PR with prefixing your classes. Please make sure they all use `rhoas-plugin` or `rhoas`.

```suggestion
.rhoas-plugin-service-instance-table tr>*:empty {
```
or
```suggestion
.rhoas-service-instance-table tr>*:empty {
```

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,92 @@
+};
+
+type PropsFromDispatch = {
+  onClickTab?: (name: string) => void;
```
```suggestion
  onClickTab: (name: string) => void;
```

wtrocki replied:
I cannot change this. onClickTab property is not passed directly but injected thru connect - which is redux state method.

christianvogt commented on frontend/packages/rhoas-plugin/src/utils/conditionHandler.ts:
```diff
@@ -0,0 +1,25 @@
+import { StatusCondition } from './rhoas-types';
+
+export const getCondition = (request: any, name: string) => {
```
Please provide a proper type for `request` instead of `any`.

christianvogt commented on frontend/packages/rhoas-plugin/src/topology/components/TopologyKafkaPanel.tsx:
```diff
@@ -0,0 +1,92 @@
+import { Alert, AlertActionCloseButton } from '@patternfly/react-core';
+import { Node } from '@patternfly/react-topology';
+import * as UIActions from '@console/internal/actions/ui';
+import './TopologyKafkaPanel.css';
```
nit: import css file as last import.

wtrocki commented on frontend/packages/rhoas-plugin/src/catalog/providers/useRhoasCatalog.tsx:
```diff
@@ -0,0 +1,143 @@
+    return cloudServicesCard;
+  }, [namespace, errorMsg, isServiceAccountValid, loaded, serviceAccount, t]);
+
+  return [services, loadedOrError, undefined];
```
This was initially there but because we have this transformation tile when CR is not there it will show constant loading screen. I spend some time trying to figure out more elegant solution, but this seems like it is the best I could come with

christianvogt replied:
@wtrocki you're right. If there's an error, you actually show a different card. In which case it isn't an error we need to present to the catalog.

Ignore my suggestion!

wtrocki commented:
> cc @vojtechszocs thoughts? and why would it cause the build to fail. Maybe we sweep all packages and remove the extension of exposed modules?

wtrocki commented:
> This comment was marked as resolved. We also discussed this on our call.
However no change has been made.

I have created a separate issue for it and hoped to add this separately, but will do that in this PR. 
Sorry for closing without comment

https://github.com/bf2fc6cc711aee1a0c2a/openshift-console-plugin/issues/140

christianvogt commented:
> > cc @vojtechszocs thoughts? and why would it cause the build to fail. Maybe we sweep all packages and remove the extension of exposed modules?
> 
> This is due to us using index.ts for catalog (so we can use folder name and second constant case refers to the exact file. There is no issue here.

I see now. My mistake as I assumed catalog was a file and didn't realize it was a directory.

christiemolloy commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceInstance.tsx:
```diff
@@ -0,0 +1,114 @@
+  disableCreateButton: boolean;
+};
+
+function areAllServicesSelected(currentServices: string[], listOfServices: CloudKafka[]) {
+  const selectedServices = listOfServices.reduce((count, service) => {
+    if (service.status !== 'ready') {
+      return count;
+    }
+    for (const selectedService of currentServices) {
+      if (selectedService === service.id) {
+        return count;
+      }
+    }
+    return count + 1;
+  }, 0);
+  return selectedServices === 0;
+}
```
Nice! haven't used `.some` before

christiemolloy commented on frontend/packages/rhoas-plugin/src/components/service-table/ServiceInstanceTable.tsx:
```diff
@@ -0,0 +1,171 @@
+              { title: name },
+              { title: provider === 'aws' ? 'Amazon Web Services' : provider },
+              { title: region },
+              { title: <a href="/">{owner}</a> },
```
Thanks! This shouldnt be linked anymore...looking at design

wtrocki replied:
Yes. let's remove it.

wtrocki commented:
@christianvogt I have addressed 

![Screenshot 2021-03-25 at 17 35 03](https://user-images.githubusercontent.com/981838/112517668-7b65ac80-8d90-11eb-824b-b5dbb5e6fb49.png)

I cannot find the original comment, got lost somewhat for me.


Also added support for not including Description in catalog panel that we talked about
Before
![Screenshot 2021-03-25 at 17 19 29](https://user-images.githubusercontent.com/981838/112517846-b7007680-8d90-11eb-8464-c9fb2bd254fe.png)
After
![Screenshot 2021-03-25 at 17 08 53](https://user-images.githubusercontent.com/981838/112517877-bff14800-8d90-11eb-823c-e61c1e89a0a9.png)

christianvogt commented:
@wtrocki the lost comment was in regards to `ServicesErrorState` and `ServicesEmptyState` being essentially the same. You should combine them into a single component.

wtrocki commented:
Applied error state unification. Adding improved version of catalog header sdk:
```
 {!details?.descriptions?.[0]?.label && (
                    <SectionHeading text={t('devconsole~Description')} />
                  )}
``

wtrocki commented:
@invincibleJai @jeff-phillips-18  I have rebased this off master. 
Our team verified this but also would be good to get your approval on this as well.

vikram-raj commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceInstance.tsx:
```diff
@@ -0,0 +1,99 @@
+/* eslint-disable prettier/prettier */
```
Is it necessary?

christianvogt replied:
good catch. we should never disable prettier

wtrocki commented on frontend/packages/topology/src/components/side-bar/TopologySideBar.tsx:
```diff
@@ -73,6 +75,9 @@ export const SelectedItemDetails: React.FC<SelectedItemDetailsProps> = observer(
       if (selectedEntity.getType() === TYPE_HELM_RELEASE) {
         return <TopologyHelmReleasePanel helmRelease={selectedEntity} />;
       }
+      if (selectedEntity.getType() === TYPE_MANAGED_KAFKA_CONNECTION) {
```
@jeff-phillips-18 I got conflict on this right now. File seems completely different. 
Any PR/Info how to rebase things

wtrocki commented:
Ready for verification. Sending cluster details using PM

vikram-raj commented:
I notice that the user has only view access to the namespace. But still able to see the `Managed Services` add card. It should not be visible for the namespace for which the user has only view access.

![image](https://user-images.githubusercontent.com/2561818/112674210-a5e86000-8e8b-11eb-9667-4daa59a960c5.png)

wtrocki commented:
@vikram-raj do you have example of other plugins that I can apply here. Is this flag feature specific?

vikram-raj commented on frontend/packages/rhoas-plugin/console-extensions.json:
```diff
@@ -0,0 +1,48 @@
+      "href": "/catalog?catalogType=managedservices",
+      "label": "%rhoas-plugin~Managed Services%",
+      "description": "%rhoas-plugin~Reduce operational complexity and focus on building and scaling applications that add more value.%",
+      "icon": { "$codeRef": "constants.cloudServicesIcon" }
```
You need to pass `accessReview` properties to check if the user has create access to this resource.
Like we did for other add cards https://github.com/openshift/console/blob/master/frontend/packages/dev-console/console-extensions.json#L10-L16

wtrocki replied:
Thank you so much @vikram-raj! Applied this suggestion and added a new item to watch out in contribution guide

wtrocki commented:
There is conflict between yarn I18n and eslint rules that put extra line break in JSON files
i18n needs to be run last or build will fail.

wtrocki commented:
@openshift/team-devconsole-ux

vikram-raj commented on frontend/packages/rhoas-plugin/src/utils/resourceCreators.ts:
```diff
@@ -0,0 +1,238 @@
+  return createCloudServicesRequest(currentNamespace);
+};
+
+export const createSecretIfNeeded = async function(
```
can we use an arrow function for consistency?

vikram-raj commented on frontend/packages/rhoas-plugin/src/topology/rhoas-topology-plugin.tsx:
```diff
@@ -0,0 +1,67 @@
+  | TopologyDataModelFactory
+  | TopologyCreateConnector;
+
+const getRhoasWatchedResources = (namespace: string): WatchK8sResources<any> => {
```
Better to have type instead of any. Or can we use `WatchK8sResource`  as we are fetching only one resource here?
```suggestion
const getRhoasWatchedResources = (namespace: string): WatchK8sResources<{ kafkaConnections: K8sResourceKind[] }> => {
```

vikram-raj commented on frontend/packages/rhoas-plugin/src/topology/components/ResourceComponent.tsx:
```diff
@@ -0,0 +1,31 @@
+import { ResourceLink } from '@console/internal/components/utils';
+import { referenceForModel } from '@console/internal/module/k8s';
+import { SecretModel } from '@console/internal/models';
+import * as React from 'react';
+import { useTranslation } from 'react-i18next';
```
nit: sort the import.

wtrocki replied:
What this means?

vikram-raj commented on frontend/packages/rhoas-plugin/src/topology/components/DetailsComponent.tsx:
```diff
@@ -0,0 +1,36 @@
+import * as React from 'react';
+import { ResourceSummary } from '@console/internal/components/utils';
+import { useTranslation } from 'react-i18next';
```
nit: sort the import.

vikram-raj commented on frontend/packages/rhoas-plugin/src/components/states/ServicesErrorState.tsx:
```diff
@@ -0,0 +1,16 @@
+      message={message}
+      actionLabel={actionLabel}
+      icon={ErrorCircle}
+      iconClass={'rhoas-services-error-state-icon'}
```
nit: 
```suggestion
      iconClass="rhoas-services-error-state-icon"
```

vikram-raj commented on frontend/packages/rhoas-plugin/src/components/service-table/ServiceInstanceTable.tsx:
```diff
@@ -0,0 +1,169 @@
+} from '@patternfly/react-table';
+import SearchIcon from '@patternfly/react-icons/dist/js/icons/search-icon';
+import { Timestamp } from '@console/internal/components/utils';
+import './ServiceInstanceTable.css';
```
nit:  sort the import. Style should import at last.  and can we use `scss` instead of css?

wtrocki replied:
Ah.. lesson learned. Lint issues rearranged that for me. I need to separate those if I need them on the bottom

vikram-raj commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceListPage.tsx:
```diff
@@ -0,0 +1,126 @@
+  getFinishedCondition,
+} from '../../utils/conditionHandler';
+import { ServicesErrorState } from '../states/ServicesErrorState';
+import { useTranslation } from 'react-i18next';
```
nit: sort the import.

vikram-raj commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceInstance.tsx:
```diff
@@ -0,0 +1,98 @@
+import ServiceInstanceTable from '../service-table/ServiceInstanceTable';
+import { ServicesEmptyState } from '../states/ServicesEmptyState';
+import { CloudKafka } from '../../utils/rhoas-types';
+import CubesIcon from '@patternfly/react-icons/dist/js/icons/cubes-icon';
```
nit: sort the import.

vikram-raj commented on frontend/packages/rhoas-plugin/locales/OWNERS:
```diff
@@ -0,0 +1,2 @@
+labels:
+  - kind/i18n
```
Can we add label for package `component/rhoas-plugin`?
```suggestion
  - kind/i18n
  - component/rhoas-plugin
```

wtrocki commented:
@vikram-raj Appling fixes right now

wtrocki commented:
@vikram-raj Amazing comment. I actually found this extra bug that is now addressed in fc15a11 along with small improvements suggested

wtrocki commented:
I also found one issue that were pointed out but missed during review. Resoling that now.

wtrocki commented:
@vikram-raj Amazing review. I actually found this extra bug that is now addressed in 
61c393d along with small improvements suggested

vikram-raj commented on frontend/packages/rhoas-plugin/src/utils/conditionHandler.ts:
```diff
@@ -0,0 +1,26 @@
+
+export const getCondition = (request: K8sResourceKind, name: string) => {
+  if (request && request.status && request.status.conditions) {
+    for (const condition of request.status.conditions) {
+      if (condition.type === name) {
+        return condition as StatusCondition;
+      }
+    }
```
I think we can make use of filter func here.

wtrocki replied:
yes. This is typical find method use case

vikram-raj commented on frontend/packages/rhoas-plugin/src/utils/conditionHandler.ts:
```diff
@@ -0,0 +1,26 @@
+
+export const isResourceStatusSuccessfull = (request: K8sResourceKind) => {
+  const condition = getCondition(request, 'Finished');
+  return condition && condition.status === 'True';
```
```suggestion
  return condition?.status === 'True';
```

vikram-raj commented on frontend/packages/rhoas-plugin/src/utils/conditionHandler.ts:
```diff
@@ -0,0 +1,26 @@
+  return undefined;
+};
+
+export const getFinishedCondition = (request: K8sResourceKind) => {
+  return getCondition(request, 'Finished');
+};
```
```suggestion
export const getFinishedCondition = (request: K8sResourceKind) => getCondition(request, 'Finished');
```

vikram-raj commented on frontend/packages/rhoas-plugin/src/utils/conditionHandler.ts:
```diff
@@ -0,0 +1,26 @@
+  return condition && condition.status === 'True';
+};
+
+export const isAcccesTokenSecretValid = (request: K8sResourceKind) => {
+  return getCondition(request, 'AcccesTokenSecretValid')?.status === 'True';
+};
```
```suggestion
export const isAcccesTokenSecretValid = (request: K8sResourceKind) =>
  getCondition(request, 'AcccesTokenSecretValid')?.status === 'True';
```

vikram-raj commented on frontend/packages/rhoas-plugin/src/topology/rhoas-data-transformer.ts:
```diff
@@ -0,0 +1,115 @@
+
+export const getTopologyRhoasNodes = (kafkaConnections: K8sResourceKind[]): NodeModel[] => {
+  const nodes = [];
+  for (const obj of kafkaConnections) {
```
I think we can make use of the `reduce` func here.

wtrocki replied:
I would like to keep it the same as other plugins like helm.

vikram-raj commented on frontend/packages/rhoas-plugin/src/topology/rhoas-data-transformer.ts:
```diff
@@ -0,0 +1,115 @@
+  if (rhoasDataModel.nodes?.length) {
+    workloads.forEach((dc) => {
+      rhoasDataModel.edges.push(
+        ...[...getRhoasServiceBindingEdges(dc, rhoasDataModel.nodes, serviceBindingRequests)],
```
```suggestion
       ...getRhoasServiceBindingEdges(dc, rhoasDataModel.nodes, serviceBindingRequests),
```
as `getRhoasServiceBindingEdges` is an array.

vikram-raj commented on frontend/packages/rhoas-plugin/src/topology/rhoas-data-transformer.ts:
```diff
@@ -0,0 +1,115 @@
+  return edges;
+};
+export const getRhoasTopologyDataModel = () => (
+  namespace: string,
```
do we need this `namespace` param as it is not been used?

wtrocki replied:
```
 Types of parameters 'resources' and 'namespace' are incompatible.
        Type 'string' is not assignable to type 
```

vikram-raj reviewed (commented):
Sorry, added few more comments that I missed.

wtrocki commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceInstance.tsx:
```diff
@@ -0,0 +1,97 @@
+import * as React from 'react';
+import { useTranslation } from 'react-i18next';
+import { FormFooter, FormHeader, FlexForm, FormBody } from '@console/shared';
+import CubesIcon from '@patternfly/react-icons/dist/js/icons/cubes-icon';
+import FormSection from '@console/dev-console/src/components/import/section/FormSection';
+import ServiceInstanceFilter from '../service-table/ServiceInstanceFilter';
+import ServiceInstanceTable from '../service-table/ServiceInstanceTable';
+import { history } from '@console/internal/components/utils';
```
I would go and create issue to configure https://eslint.org/docs/rules/sort-imports 
Currently lint sorts imports for me but not exactly like suggested above.

Happy to apply this changes etc. but best to resolve this by automation in the future.

CC @christianvogt

wtrocki commented on frontend/packages/rhoas-plugin/src/components/states/ServicesErrorState.tsx:
```diff
@@ -0,0 +1,16 @@
+import './ServicesErrorState.css';
+import { ServicesEmptyState, ServicesEmptyStateProps } from './ServicesEmptyState';
+
+export const ServicesErrorState = ({ title, message, actionLabel }: ServicesEmptyStateProps) => {
+  return (
+    <ServicesEmptyState
+      title={title}
+      message={message}
+      actionLabel={actionLabel}
+      icon={ErrorCircle}
+      iconClass="rhoas-services-error-state-icon"
+    />
+  );
+};
```
Yes. I'm old school dev :)

wtrocki commented on frontend/packages/rhoas-plugin/src/utils/resourceCreators.ts:
```diff
@@ -0,0 +1,235 @@
+  k8sUpdate,
+  k8sKillByName,
+} from '@console/internal/module/k8s/resource';
+
```
eslint doesn't let me :D

wtrocki commented on frontend/packages/rhoas-plugin/src/utils/conditionHandler.ts:
```diff
@@ -0,0 +1,26 @@
+import { StatusCondition } from './rhoas-types';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+
```
eslint doesn't let me do it

wtrocki commented on frontend/packages/rhoas-plugin/src/utils/resourceCreators.ts:
```diff
@@ -0,0 +1,235 @@
+ * @param kafkaName
+ * @param currentNamespace
+ */
+export const deleteKafkaConnection = (kafkaName: string, currentNamespace: string) => {
+  return k8sKillByName(KafkaConnectionModel, kafkaName, currentNamespace);
+};
```
not fan of this style but would do it :)

wtrocki commented on frontend/packages/rhoas-plugin/src/components/states/ServicesErrorState.tsx:
```diff
@@ -0,0 +1,16 @@
+import './ServicesErrorState.css';
+import { ServicesEmptyState, ServicesEmptyStateProps } from './ServicesEmptyState';
+
+export const ServicesErrorState = ({ title, message, actionLabel }: ServicesEmptyStateProps) => {
```
Not 100% sure about this but done

christianvogt replied:
We should actually be typing components with `React.FC` instead of simply typing the param.

wtrocki commented on frontend/packages/rhoas-plugin/src/components/states/ServicesEmptyState.tsx:
```diff
@@ -0,0 +1,36 @@
+  iconClass?: string;
+};
+
+export const ServicesEmptyState = ({
+  title,
+  message,
+  actionLabel,
+  icon,
+  iconClass,
+}: ServicesEmptyStateProps) => {
+  return (
+    <EmptyState>
+      <EmptyStateIcon className={iconClass} icon={icon} />
+      <Title headingLevel="h4" size="lg">
+        {title}
+      </Title>
+      {message || (
+        <Title headingLevel="h5" size="md">
+          {message}
+        </Title>
+      )}
+      <Button variant="link" onClick={history.goBack}>
+        {actionLabel}
+      </Button>
+    </EmptyState>
+  );
+};
```
I really dislike this style and there is no extra benefit in performance or resulting component. Is there any written rule how components should be done or something?

christianvogt replied:
There's really no difference. TBH if we want to enforce one style over another, we should create an eslint rule.

wtrocki commented:
@vikram-raj @invincibleJai @christianvogt PR is ready for review.

christianvogt commented on frontend/packages/rhoas-plugin/src/components/states/ServicesErrorState.tsx:
```diff
@@ -0,0 +1,16 @@
+import * as React from 'react';
+import ErrorCircle from '@patternfly/react-icons/dist/js/icons/error-circle-o-icon';
+import './ServicesErrorState.css';
+import { ServicesEmptyState, ServicesEmptyStateProps } from './ServicesEmptyState';
```
Agreed. We should use `React.ComponentProps<typeof ServicesEmptyState>` instead of exporting the prop types.

christianvogt commented on frontend/packages/rhoas-plugin/src/utils/resourceCreators.ts:
```diff
@@ -0,0 +1,235 @@
+  CloudServicesRequestModel,
+} from '../models/rhoas';
+import { getFinishedCondition, isResourceStatusSuccessfull } from './conditionHandler';
+import { k8sWaitForUpdate } from '@console/internal/module/k8s';
```
eslint import rule has been disabled. We shouldn't nit on imports, but instead take the time to fix the linter so that it's auto fixed.

wtrocki commented:
During offline review some issues were pointed. I have addressed them :D

vikram-raj commented:
Thanks @wtrocki, Varified the changes and it works as expected.

/lgtm

wtrocki commented:
/assign @christianvogt

invincibleJai commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceListPage.tsx:
```diff
@@ -0,0 +1,124 @@
+
+  React.useEffect(() => {
+    const createKafkaRequestFlow = async () => {
+      await createCloudServicesRequestIfNeeded(currentNamespace);
+
+      const currentKafka = await listOfCurrentKafkaConnectionsById(currentNamespace);
```
is it safe with `await` here or we would need to wrap in `try - catch` ?

invincibleJai commented on frontend/packages/rhoas-plugin/src/components/service-list/ServiceListPage.tsx:
```diff
@@ -0,0 +1,124 @@
+
+  const createKafkaConnectionFlow = React.useCallback(async () => {
+    setSubmitting(true);
+    const { id, name } = remoteKafkaInstances[selectedKafka];
```
this may break if `remoteKafkaInstances` is `undefined`? good to have check or rely on `loaded && !errMessage` of `useK8sWatchResource` above. WDYT

wtrocki replied:
At this point this never will be undefined. Also when original issue of flickering was pointed out today I have mentioned that this is temporary solution and we have plans to rewrite listPate/tables.

invincibleJai replied:
Thanks for input m have felt as in line number 55 we have as check as below

```
const remoteKafkaInstances = watchedKafkaRequest?.status?.userKafkas;
```

if we can make provide default then will be safe

```
const remoteKafkaInstances = watchedKafkaRequest?.status?.userKafkas ?? { }
```

invincibleJai commented:
/lgtm

Thanks @wtrocki, Have verified the changes locally, and overall looks good 🎉 

One observation that can be covered in later PRs, as well as the design, doesn't capture it
- Actions are shown in the context-menu on `KC` should show up in the side-panel as well and the title should be clickable.

invincibleJai commented:
/lgtm

christianvogt commented:
> /lgtm
> 
> Thanks @wtrocki, Have verified the changes locally, and overall looks good 🎉
> 
> One observation that can be covered in later PRs, as well as the design, doesn't capture it
> 
> * Actions are shown in the context-menu on `KC` should show up in the side-panel as well and the title should be clickable.

@wtrocki please create a jira issue to fix this afterwards

christianvogt commented:
/approve

wtrocki commented:
/test e2e-gcp-console

wtrocki commented:
> please create a jira issue to fix this afterwards

I have this logged in GitHub and will do it in second PR
